### ISSUE TITLE
Bug: [no-floating-promises] any-typed rejection handler misclassified as 'not a function'

### ISSUE THREAD

**Issue by martin-strecker-sonarsource** [typescript-eslint/typescript-eslint#12848]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have searched for related issues and found none that matched my issue (closest are #9808 "should not be tricked by spread args" and #7135 "return type of .catch not checked" — both are different mechanisms).
- [x] I have read the FAQ and my problem is not listed.

### Playground Link

No playground link — repro requires `allowJs`/no strict lib config, reproduced locally with `RuleTester` against `@typescript-eslint/eslint-plugin` directly. Repro code below.

### Repro Code

Plain JS, untyped parameter (most common real-world shape — e.g. Node-style callback-style code):

```js
// tsconfig: { "allowJs": true }
function run(next) {
  Promise.reject(new Error('boom')).catch(next);
}
```

Equivalent with an explicit `any` annotation in a `.ts` file, showing this isn't a JS-inference quirk:

```ts
function run(next: any) {
  Promise.reject(new Error('boom')).catch(next);
}
```

Both are reported by `no-floating-promises` with:

```
Promises must be awaited, end with a call to .catch, end with a call to .then with a
rejection handler or be explicitly marked as ignored with the `void` operator. A rejection
handler that is not a function will be ignored.
```
(`messageId: floatingUselessRejectionHandlerVoid`)

Control case — a properly-typed handler is correctly **not** reported, confirming the bug is specific to `any`, not to passing a handler by identifier in general:

```ts
function run(next: (err: Error) => void) {
  Promise.reject(new Error('boom')).catch(next); // correctly not reported
}
```

### Expected Result

None of the two `any`-typed cases above should be reported — the handler genuinely is a function at runtime in all three cases (it's only the *declared type* that differs).

### Actual Result

Both `any`-typed cases are reported as if the handler weren't callable at all.

### Additional Info

Root cause is in `isValidRejectionHandler` (`packages/eslint-plugin/src/rules/no-floating-promises.ts`):

```ts
function isValidRejectionHandler(rejectionHandler: TSESTree.Node): boolean {
  return (
    services.program
      .getTypeChecker()
      .getTypeAtLocation(services.esTreeNodeToTSNodeMap.get(rejectionHandler))
      .getCallSignatures().length > 0
  );
}
```

`getCallSignatures()` structurally returns an empty array for a value typed `any` — TypeScript doesn't consider `any` to "have" call signatures, even though `any` is callable at runtime by definition (that's the whole point of `any`). So any legitimate rejection handler that resolves to `any` — an untyped JS callback parameter, an explicit `: any` annotation, or a global whose declaration isn't in scope due to missing `lib`/`types` (e.g. `console.error` used as a `.catch()` handler in a project whose `tsconfig.json` doesn't include `"node"`/DOM lib) — gets misclassified as non-callable.

We hit this for real in three different open-source projects while validating a new Sonar rule built on top of `no-floating-promises` (791 findings reviewed manually): twice for an untyped Mocha-style callback parameter, once for a rest-args-reassigned untyped parameter, and once for `console.error` used as a `.catch()` handler in a project missing `"node"` lib types.

### Suggested Fix

`ts-api-utils` (already imported as `tsutils` in this file) exports `isIntrinsicAnyType`. Short-circuiting on it before the call-signature check resolves all four real-world false positives above without weakening genuine detection (verified locally: a concretely-typed non-function handler, e.g. a `string`-typed value, is still correctly flagged):

```ts
function isValidRejectionHandler(rejectionHandler: TSESTree.Node): boolean {
  const type = services.program
    .getTypeChecker()
    .getTypeAtLocation(services.esTreeNodeToTSNodeMap.get(rejectionHandler));
  // `any` structurally has zero call signatures even though it's callable at
  // runtime by definition (untyped JS parameter, explicit `: any`, or a
  // global missing from `lib`/`types`). Assume it's valid rather than
  // reporting a near-certain false positive.
  if (tsutils.isIntrinsicAnyType(type)) {
    return true;
  }
  return type.getCallSignatures().length > 0;
}
```

I'm happy to open a PR with this change (plus tests) if that's welcome.

**Comment by bradzacher** [comment]
Isn't this correct to report? `any` is very potentially NOT a function.

**Comment by kirkwaiblinger** [comment]
@martin-strecker-sonarsource 

What happened to this part of the issue template?

> - [ ] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

**Comment by martin-strecker-sonarsource** [comment]
>Isn't this correct to report? any is very potentially NOT a function.

When passed to `catch` it likely is a function.

The message explicitly says "A rejection handler that is not a function will be ignored." That's why `isValidRejectionHandler` tests "is this parameter callable" in the first place. For an `any`-typed value, the honest answer is "we don't know" and that uncertainty resolves to either FPs (current behavior) or FNs (what's proposed here). You call which weighs more.

One data point that might cut either way: the same rule already has to make this same any-shaped call elsewhere. https://github.com/typescript-eslint/typescript-eslint/blob/cd2e3d796b5e18cdaa4ea10fec0b87caaa8db026/packages/eslint-plugin/src/rules/no-floating-promises.ts#L356

`isPromiseLike` doesn't treat an `any`-typed expression as promise-like at all, so a floating `any` is never reported by this rule in the first place. This asymmetry looks more like an oversight than a deliberate stance. And even if we make `isValidRejectionHandler` consistent with `isPromiseLike` here, the case that flips from FP to FN is still independently caught by `no-unsafe-argument`, since `.catch()`'s parameter is typed as a function, not `any`. So the change costs nothing in practice.

**Comment by martin-strecker-sonarsource** [comment]
> What happened to this part of the issue template?

Sorry. You are right. This was an AI authored issue, but the fix was probe tested as part of this PR https://github.com/SonarSource/SonarJS/pull/7896

I will be more careful next time when I open the issue and I will not remove that checkbox from the template.

**Comment by bradzacher** [comment]
> isPromiseLike doesn't treat an any-typed expression as promise-like at all, so a floating any is never reported by this rule in the first place.

Yes, because otherwise we'd report on every single expression statement that has an `any` type - which is way too broad.

In this instance were specifically looking at arguments passed to a known promises `.catch` handler - so the scope is very, very targeted.

IMO reporting on `any` is fine here - the chance of false postives and false negatives are both well defined and the chance of reporting a case that could be a production bug is high.

This is also easily fixed from the user side:
- a disable comment mentioning that it is a false negative and that the `any` is known to be a function (worst choice)
- an assertion to change the type to a function type to clearly document and encode the expectation
- an annotation on the parameter/variable to set its type to a function type (even just the `Function` type) to clearly encode the expectation.

All of these are very reasonable choices for this report.

For these reasons I'm going to close this as working as intended.
